### PR TITLE
injector: add new injectedAnnotations to support PSP

### LINKED ISSUE #17331: Extend injection template to support injecting custom annotations
**Bug description**

Istio's sidecar injection webhook rewrites the pod spec to inject the Istio containers, _after_ the pod spec has been rewritten by the PodSecurityPolicy admission controller to match the applied PSP, but also _before_ the PSP validation (see https://github.com/istio/istio/issues/17318 for more details). In the case the spec injected by Istio would need to be mutated by the PSP admission controller to add annotations specific to the injected containers, the PSP validation would reject the resulting pod spec, and the pods couldn't be created. In the case of a `Deployment`, the `ReplicaSet` description would show an error like:

```
Error creating: pods "httpgo-796d5fdfbf-xhz2x" is forbidden: unable to validate against any pod security policy: []
```

For example, if the matching PSP contains the following apparmor-specific annotations:
```
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  annotations:
    apparmor.security.beta.kubernetes.io/allowedProfileNames: runtime/default
    apparmor.security.beta.kubernetes.io/defaultProfileName: runtime/default
```

The PSP admission controller may add the following annotation for each container in the pod spec:
```
container.apparmor.security.beta.kubernetes.io/<container name>: runtime/default
```

If the Istio sidecar injection is done after the PSP admission control, as is the case when using the automatic injection webhook, the pod spec will be missing the apparmor annotations for the Istio containers:
```
container.apparmor.security.beta.kubernetes.io/istio-init: runtime/default
container.apparmor.security.beta.kubernetes.io/istio-proxy: runtime/default
```

As a consequence, the resulting pod spec will be rejected by PSP validation.

**This issue is about allowing users to specify additional annotations to be added by Istio injection, e.g. the missing apparmor annotations above.** The injection template should be extended to support specifying those additional annotations.

**Affected product area (please put an X in all that apply)**

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

All Istio versions.

**Environment where bug was observed (cloud vendor, OS, etc)**

Any cluster with pod security policy enabled.

### PR DESCRIPTION
**Pull request by howardjohn**
Add a new injectedAnnotations setting for the injector. Any annotations
here will be added to the final pod spec. This is required for PSPs to
work. See https://github.com/istio/istio/issues/17331 for details.

Note - a lot of changes here are to existing tests just to allow setting
all config options on the webhook rather than just the template.

Fixes https://github.com/istio/istio/issues/17331

### CODE REVIEW

File: pkg/kube/inject/webhook.go [hunk 1]
```diff
@@ -460,7 +461,16 @@ func addLabels(target map[string]string, added map[string]string) []rfc6902Patch
 }
 
 func updateAnnotation(target map[string]string, added map[string]string) (patch []rfc6902PatchOperation) {
-	for key, value := range added {
+	log.Errorf("howardjohn: added %v", added)
```
Comment (morvencao) [inline pkg/kube/inject/webhook.go]: remove this debugging message?

Comment (rshriram) [reply pkg/kube/inject/webhook.go]: this is his secret plan for becoming famous and getting his name printed on all the elasticsearches and splunks in the planet :P

Comment (howardjohn) [reply pkg/kube/inject/webhook.go]: You caught me... just reverted

**Review by morvencao — APPROVED** [review]

**Comment by howardjohn** [comment]
@rshriram mind approving the changes to injector? (we should probably add an owner to this)

**Review by rshriram — APPROVED** [review]

**Comment by thecodejunkie** [comment]
@howardjohn 

I can't for the life of me figure out the syntax for passing in values to the **sidecarInjectorWebhook.injectedAnnotations** parameter, to set the apparmor annotation 🙂My issue is that the "key" part contains dot-notation so it attempts to split it up no matter how I try and encapsulate it

Using Helm I think the proper way is to encapsulate it in quotes and escape dots
```
istioctl manifest generate \
    --set cni.enabled=true \
    --set values.sidecarInjectorWebhook.injectedAnnotations."container\.apparmor\.security\.beta\.kubernetes\.io/istio-proxy"=runtime/default \
    --output ./generated
```

However this results in a mess in the actual manifest

```
injectedAnnotations:
  "container\": "map[apparmor\:map[security\:map[beta\:map[kubernetes\:map[io/istio-proxy:runtime/default]]]]]"
```

Thanks!

**Comment by howardjohn** [comment]
@thecodejunkie adding complex stuff like this is really painful with `helm --set` and may not even be possible with istioctl.. I am not sure. What I would recommend is just to use a file. This will work:

```
ik manifest generate -f <(echo 'apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  values:
    sidecarInjectorWebhook:
      injectedAnnotations:
        container.apparmor.security.beta.kubernetes.io/istio-proxy: runtime/default')
```

@richardwxn is this the expected behavior?

**Comment by thecodejunkie** [comment]
@howardjohn thanks for the update! I believe setting it with `helm --set` should work if you use the syntax that I used above https://stackoverflow.com/a/50399612 but it might be that `istioctl` incorrectly (?) handles the escaping? You offered solution definitly works as it saves me from having to manually edit the configmap afterwards (eww to manual steps in a setup process 😄)

**Comment by howardjohn** [comment]
Yeah the istioctl set is not as complete as helm, might not be supported. Let me open an issue for this so we don't loose the info here

**Comment by thecodejunkie** [comment]
@howardjohn awesome! Thanks

What would be the proper way to move the `cni.enabled=true` config into that file? Straight under `values`?

```
k manifest generate -f <(echo 'apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  values:
    cni:
      enabled: true
    sidecarInjectorWebhook:
      injectedAnnotations:
        container.apparmor.security.beta.kubernetes.io/istio-proxy: runtime/default')
```

**Comment by howardjohn** [comment]
@thecodejunkie youll want to do

```
k manifest generate -f <(echo 'apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  cni:
    enabled: true
  values:
    sidecarInjectorWebhook:
      injectedAnnotations:
        container.apparmor.security.beta.kubernetes.io/istio-proxy: runtime/default')
```


The --set path is basically just a flattened version of the ICP CRD. It is admittedly a bit confusing where things should be under `values` and when things shouldn't, we are still iterating on this and trying to improve

**Comment by thecodejunkie** [comment]
Thanks! I guess one way to think about it is "if it's in a sub-chart" then it goes under `values`

**Pull request merged**